File: 0098-validate-binary-search-tree/0098-validate-binary-search-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
private:
    bool dfs(TreeNode* root, long mini, long maxi) {
        // If the current node is null, it is a valid subtree
        if (!root) {
            return true;
        }

        // Check if the current node's value is within the valid range (mini, maxi)
        if (root->val <= mini || root->val >= maxi) {
            return false;
        }

        // Recursively check the left and right subtrees
        // Update the range for the left subtree to (mini, root->val)
        // Update the range for the right subtree to (root->val, maxi)
        return dfs(root->left, mini, root->val) && dfs(root->right, root->val, maxi);
    }

public:
    bool isValidBST(TreeNode* root) {
        // Start the DFS with an initial valid range for values in the tree
        return dfs(root, LONG_MIN, LONG_MAX);
    }
};

File: 0098-validate-binary-search-tree/0098-validate-binary-search-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private boolean dfs(TreeNode node, long minRange, long maxRange) {
        if(node == null) return true;
        if(node.val <= minRange || node.val >= maxRange) {
            return false;
        }
        return dfs(node.left, minRange, node.val) && dfs(node.right, node.val, maxRange);
    }

    public boolean isValidBST(TreeNode root) {
        return dfs(root, Long.MIN_VALUE, Long.MAX_VALUE);
    }
}